plugin/trino-elasticsearch/src/main/java/io/trino/plugin/elasticsearch/ScanQueryPageSource.java

```diff
@@ -267,67 +260,75 @@ private void flattenFields(Map<String, Type> result, String fieldName, Type type
                         return new SourceColumnDecoder();
                     }
 
-                    return createDecoder(column.getName(), column.getType());
+                    return createDecoder(
+                            column.getName(),
+                            column.getType(),
+                            column.getColumnProperty());
                 })
                 .collect(toImmutableList());
     }
 
-    private Decoder createDecoder(String path, Type type)
+    private Decoder createDecoder(String path, Type type, ElasticsearchColumnProperty columnProperty)
     {
-        if (type.equals(VARCHAR)) {
-            return new VarcharDecoder(path);
+        if (columnProperty == null) {
+            System.out.println("hi");
         }
-        if (type.equals(VARBINARY)) {
-            return new VarbinaryDecoder(path);
+        //TODO remove type field dependency here (ip address, row, and array types)
+        switch (columnProperty.getDecoderType()) {
+            case RAW_JSON:
+                return new RawJsonDecoder(path);
+            case VARCHAR:
+                return new VarcharDecoder(path);
+            case VARBINARY:
+                return new VarbinaryDecoder(path);
+            case TIMESTAMP:
+                return new TimestampDecoder(path);
+            case BOOLEAN:
+                return new BooleanDecoder(path);
+            case DOUBLE:
+                return new DoubleDecoder(path);
+            case REAL:
+                return new RealDecoder(path);
+            case TINYINT:
+                return new TinyintDecoder(path);
+            case SMALLINT:
+                return new SmallintDecoder(path);
+            case INTEGER:
+                return new IntegerDecoder(path);
+            case BIGINT:
+                return new BigintDecoder(path);
+            case IP_ADDRESS:
+                return new IpAddressDecoder(path, type);
+            case ROW:
```

Comment:
Row decoder should be able to know the order of children from `DecoderContext` (i.e. `ElasticSearchColumnProperty` in the current PR). It can just store an ordered list of children (unless we require lookups at some point). 

Other than that, type parameter should only be used for generating IpAddress decoder. We can separately remove the dependency on that by injecting TypeManager in the ScanQueryPageSource.